javaharness/java/arcs/api/PecPortManager.java

```diff
@@ -8,49 +8,71 @@
 @Singleton
 public class PecPortManager {
 
-  private final Map<String, PecMessageReceiver> ports = new HashMap<>();
-  private final PecInnerPortFactory portFactory;
+  private final ArcsMessageSender arcsMessageSender;
+  private final ParticleExecutionContext pec;
+  private final PortableJsonParser jsonParser;
+
+  private final Map<String, PecInnerPort> pecPortMap = new HashMap<>();
+  private final Map<String, PecInnerPortProxy> pecPortProxyMap = new HashMap<>();
 
   @Inject
-  PecPortManager(PecInnerPortFactory portFactory) {
-    this.portFactory = portFactory;
+  PecPortManager(
+    ArcsMessageSender arcsMessageSender,
+    ParticleExecutionContext pec,
+    PortableJsonParser jsonParser) {
+    this.arcsMessageSender = arcsMessageSender;
+    this.pec = pec;
+    this.jsonParser = jsonParser;
   }
 
   public void deliverPecMessage(String pecId, String sessionId, PortableJson message) {
-    PecMessageReceiver port = ports.get(pecId);
-    if (port == null) {
-      // Create a new inner PEC port for this pecId.
-      port = createInnerPecPort(pecId, sessionId);
+    PecInnerPort port = pecPortMap.get(pecId);
+    if (port != null) {
+      port.onReceivePecMessage(message);
+      return;
+    }
+
+    PecInnerPortProxy portProxy = pecPortProxyMap.get(pecId);
+    if (portProxy != null) {
+      portProxy.onReceivePecMessage(message);
+      return;
     }
 
+    // Create a new inner PEC port for this pecId.
+    port = createPecInnerPort(pecId, sessionId);
     port.onReceivePecMessage(message);
   }
 
-  public void addRemotePecPort(String pecId, RemotePecPort remotePecPort) {
-    ports.put(pecId, remotePecPort);
+  public void addPecInnerPortProxy(String pecId, PecInnerPortProxy pecInnerPortProxy) {
+    pecPortProxyMap.put(pecId, pecInnerPortProxy);
   }
 
-  public PecInnerPort getOrCreateInnerPort(String pecId, String sessionId) {
-    PecMessageReceiver port = ports.get(pecId);
+  public PecInnerPort getOrCreatePecInnerPort(String pecId, String sessionId) {
+    PecInnerPort port = pecPortMap.get(pecId);
     if (port == null) {
-      return createInnerPecPort(pecId, sessionId);
-    } else if (port instanceof PecInnerPort) {
-      return (PecInnerPort) port;
+      return createPecInnerPort(pecId, sessionId);
     } else {
-      throw new IllegalArgumentException("PEC with ID %s is not an inner port: " + pecId);
+      return port;
     }
   }
 
   public void removePecPort(String pecId) {
-    ports.remove(pecId);
+    if (pecPortMap.remove(pecId) != null) {
```

Comment:
can you, please, add a TODO here for me to split the method and verify it is called everywhere it needs to be called. thanks.

Reply:
Done.